File: android/versioned-abis/expoview-abi37_0_0/src/main/java/abi37_0_0/expo/modules/sensors/services/BaseService.java
// Copyright 2015-present 650 Industries. All rights reserved.

package abi37_0_0.expo.modules.sensors.services;

import android.content.Context;

import abi37_0_0.org.unimodules.core.ModuleRegistry;
import abi37_0_0.org.unimodules.core.interfaces.RegistryLifecycleListener;
import abi37_0_0.org.unimodules.core.interfaces.LifecycleEventListener;
import abi37_0_0.org.unimodules.core.interfaces.services.UIManager;

/* package */ abstract class BaseService implements LifecycleEventListener, RegistryLifecycleListener {
  private Context mContext;
  private ModuleRegistry mModuleRegistry;
  private boolean mIsForegrounded = false;

  /* package */ BaseService(Context context) {
    mContext = context;
  }

  protected Context getContext() {
    return mContext;
  }

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;

    // Register to new UIManager
    if (mModuleRegistry != null && mModuleRegistry.getModule(UIManager.class) != null) {
      mModuleRegistry.getModule(UIManager.class).registerLifecycleEventListener(this);
    }
  }

  @Override
  public void onDestroy() {
    // Unregister from old UIManager
    if (mModuleRegistry != null && mModuleRegistry.getModule(UIManager.class) != null) {
      mModuleRegistry.getModule(UIManager.class).unregisterLifecycleEventListener(this);
    }
  }

  @Override
  public void onHostResume() {
    mIsForegrounded = true;
    onExperienceForegrounded();
  }

  @Override
  public void onHostDestroy() {
    // do nothing
  }

  @Override
  public void onHostPause() {
    mIsForegrounded = false;
    onExperienceBackgrounded();
  }

  /* package */ boolean getExperienceIsForegrounded() {
    return mIsForegrounded;
  }

  abstract public void onExperienceForegrounded();
  abstract public void onExperienceBackgrounded();
}

File: android/versioned-abis/expoview-abi47_0_0/src/main/java/abi47_0_0/expo/modules/sensors/services/BaseService.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package abi47_0_0.expo.modules.sensors.services

import android.content.Context
import abi47_0_0.expo.modules.core.ModuleRegistry
import abi47_0_0.expo.modules.core.interfaces.LifecycleEventListener
import abi47_0_0.expo.modules.core.interfaces.RegistryLifecycleListener
import abi47_0_0.expo.modules.core.interfaces.services.UIManager

/* internal */
abstract class BaseService(protected val context: Context) : LifecycleEventListener, RegistryLifecycleListener {
  private lateinit var mModuleRegistry: ModuleRegistry
  var experienceIsForegrounded = false
    private set

  override fun onCreate(moduleRegistry: ModuleRegistry) {
    mModuleRegistry = moduleRegistry

    // Register to new UIManager
    mModuleRegistry.getModule(UIManager::class.java)?.registerLifecycleEventListener(this)
  }

  override fun onDestroy() {
    // Unregister from old UIManager
    if (this::mModuleRegistry.isInitialized) {
      mModuleRegistry.getModule(UIManager::class.java)?.unregisterLifecycleEventListener(this)
    }
  }

  override fun onHostResume() {
    experienceIsForegrounded = true
    onExperienceForegrounded()
  }

  override fun onHostDestroy() = Unit

  override fun onHostPause() {
    experienceIsForegrounded = false
    onExperienceBackgrounded()
  }

  abstract fun onExperienceForegrounded()
  abstract fun onExperienceBackgrounded()
}
